gemini/agent-engine/tutorial_plangen_on_naturalplan_examples.ipynb

```diff
@@ -0,0 +1,2325 @@
+        "            model_name=self.model_name\n",
+        "        )\n",
+        "\n",
+        "        # Parse reward and ensure it's an integer between -100 and 100\n",
+        "        try:\n",
+        "            reward = int(reward_output.split('Score:')[-1].strip())\n",
+        "            reward = max(-100, min(100, reward))  # Clamp between -100 and 100\n",
+        "        except:\n",
+        "            reward = -100\n",
+        "\n",
+        "        return (output, reward_output, reward)\n",
+        "\n",
+        "    def get_best_response(self, prompt, no_of_times=5):\n",
+        "        lst_of_resp = []\n",
+        "\n",
+        "        for i in range(no_of_times):\n",
+        "            output = call_gemini(prompt, temperature=0.7, model_name=self.model_name)\n",
+        "            out_with_reward = self.calculate_reward(prompt, output)\n",
+        "            lst_of_resp.append(out_with_reward)\n",
+        "\n",
+        "        # Sort responses by reward score and get the best one\n",
+        "        sorted_responses = sorted(lst_of_resp, key=lambda x: x[2], reverse=True)\n",
+        "        best_response = sorted_responses[0][0]\n",
+        "\n",
+        "        return sorted_responses, best_response"
+      ],
+      "metadata": {
+        "id": "qSjF16q1KSKW"
+      },
+      "execution_count": null,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "# Implementation of PlanGEN (Tree-of-Thought)\n",
+        "\n",
+        "> Tree-of-Thought (ToT) algorithm has been studied in detail for solving many complex problems. We modified the ToT algorithm with our constraint and verification agents. The method begins by initializing a root node that represents the problem and generating multiple potential next steps, creating a tree-like structure. The generated steps are verified using a verification agent which assigns reward scores based on a set of constraints. The iterative process involves evaluating all possible steps at a given depth, selecting the most promising path based on reward scores, and expanding it further by generating new steps. This process continues until a valid solution is identified or a pre-defined limit on iterations is reached."
+      ],
+      "metadata": {
+        "id": "ywcg_SarP_qX"
+      }
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "**Import necessary libraries for PlanGEN (Tree-of-Thought)**"
+      ],
+      "metadata": {
+        "id": "6qJ8QtQ9QStz"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "from __future__ import annotations\n",
+        "from typing import Tuple\n",
+        "import random\n",
+        "import math\n",
+        "from collections import deque\n",
+        "from enum import Enum\n",
+        "from pydantic import BaseModel\n",
+        "import tqdm\n",
+        "import numpy as np\n",
+        "\n",
```

Comment:
Move imports to a single cell at the top. Also be sure to install all extra libraries in the install cell. Not everyone will use colab which already has some of these installed.